// jsreport-xlsx/lib/index.js
'use strict';

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _recipe = require('./recipe.js');

var _recipe2 = _interopRequireDefault(_recipe);

var _serialize = require('./serialize.js');

var _serialize2 = _interopRequireDefault(_serialize);

var _shortid = require('shortid');

var _shortid2 = _interopRequireDefault(_shortid);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _vm = require('vm');

var _vm2 = _interopRequireDefault(_vm);

var _responseXlsx = require('./responseXlsx.js');

var _responseXlsx2 = _interopRequireDefault(_responseXlsx);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FS = _bluebird2.default.promisifyAll(_fs2.default);
var defaultXlsxTemplate = void 0;

module.exports = function (reporter, definition) {
  // used by html-to-xlsx recipe
  reporter.xlsx = { responseXlsx: _responseXlsx2.default };

  reporter.extensionsManager.recipes.push({
    name: 'xlsx',
    execute: _recipe2.default
  });

  if (reporter.compilation) {
    reporter.compilation.resource('defaultXlsxTemplate.json', _path2.default.join(__dirname, '../static', 'defaultXlsxTemplate.json'));
    reporter.compilation.resource('helpers.js', _path2.default.join(__dirname, '../static', 'helpers.js'));
  }

  reporter.options.tasks.modules.push({
    alias: 'fsproxy.js',
    path: _path2.default.join(__dirname, '../lib/fsproxy.js')
  });

  reporter.options.tasks.modules.push({
    alias: 'lodash',
    path: require.resolve('lodash')
  });

  reporter.options.tasks.modules.push({
    alias: 'xml2js',
    path: require.resolve('xml2js')
  });

  if (reporter.options.tasks.allowedModules !== '*') {
    reporter.options.tasks.allowedModules.push('path');
  }

  reporter.documentStore.registerEntityType('XlsxTemplateType', {
    _id: { type: 'Edm.String', key: true },
    'shortid': { type: 'Edm.String' },
    'name': { type: 'Edm.String', publicKey: 'true' },
    'contentRaw': { type: 'Edm.Binary', document: { extension: 'xlsx' } },
    'content': { type: 'Edm.String', document: { extension: 'txt' } }
  });

  reporter.documentStore.registerEntitySet('xlsxTemplates', {
    entityType: 'jsreport.XlsxTemplateType',
    humanReadableKey: 'shortid',
    splitIntoDirectories: true
  });

  reporter.documentStore.registerComplexType('XlsxTemplateRefType', {
    'shortid': { type: 'Edm.String' }
  });

  if (reporter.documentStore.model.entityTypes['TemplateType']) {
    reporter.documentStore.model.entityTypes['TemplateType'].xlsxTemplate = { type: 'Collection(jsreport.XlsxTemplateRefType)' };
  }

  reporter.initializeListeners.add('xlsxTemplates', function () {
    reporter.documentStore.collection('xlsxTemplates').beforeInsertListeners.add('xlsxTemplates', function (doc) {
      doc.shortid = doc.shortid || _shortid2.default.generate();
      return (0, _serialize2.default)(doc.contentRaw, reporter.options.tempDirectory).then(function (serialized) {
        return doc.content = serialized;
      });
    });

    reporter.documentStore.collection('xlsxTemplates').beforeUpdateListeners.add('xlsxTemplates', function (query, update, req) {
      if (update.$set && update.$set.contentRaw) {
        return (0, _serialize2.default)(update.$set.contentRaw, reporter.options.tempDirectory).then(function (serialized) {
          return update.$set.content = serialized;
        });
      }
    });
  });

  reporter.beforeRenderListeners.insert({ after: 'data' }, 'xlsxTemplates', function (req) {
    if (req.template.recipe !== 'xlsx') {
      return;
    }

    var findTemplate = function findTemplate() {
      if (!req.template.xlsxTemplate || !req.template.xlsxTemplate.shortid && !req.template.xlsxTemplate.content) {
        if (defaultXlsxTemplate) {
          return _bluebird2.default.resolve(defaultXlsxTemplate);
        }

        if (reporter.execution) {
          return _bluebird2.default.resolve(JSON.parse(reporter.execution.resource('defaultXlsxTemplate.json').toString()));
        }

        return FS.readFileAsync(_path2.default.join(__dirname, '../static', 'defaultXlsxTemplate.json')).then(function (content) {
          return JSON.parse(content);
        });
      }

      if (req.template.xlsxTemplate.content) {
        return (0, _serialize2.default)(req.template.xlsxTemplate.content, reporter.options.tempDirectory).then(function (serialized) {
          return JSON.parse(serialized);
        });
      }

      return reporter.documentStore.collection('xlsxTemplates').find({ shortid: req.template.xlsxTemplate.shortid }, req).then(function (docs) {
        if (!docs.length) {
          throw new Error('Unable to find xlsx template with shortid ' + req.template.xlsxTemplate.shortid);
        }

        return JSON.parse(docs[0].content);
      });
    };

    return findTemplate().then(function (t) {
      req.data = req.data || {};
      req.data.$xlsxTemplate = t;
      req.data.$xlsxModuleDirname = _path2.default.join(__dirname, '../');
      req.data.$tempDirectory = reporter.options.tempDirectory;
      req.data.$addBufferSize = definition.options.addBufferSize || 50000000;
      req.data.$escapeAmp = definition.options.escapeAmp;
      req.data.$numberOfParsedAddIterations = definition.options.numberOfParsedAddIterations == null ? 50 : definition.options.numberOfParsedAddIterations;

      return (reporter.execution ? _bluebird2.default.resolve(reporter.execution.resource('helpers.js')) : FS.readFileAsync(_path2.default.join(__dirname, '../', 'static', 'helpers.js'), 'utf8')).then(function (content) {
        if (req.template.helpers && _typeof(req.template.helpers) === 'object') {
          // this is the case when the jsreport is used with in-process strategy
          // and additinal helpers are passed as object
          // in this case we need to merge in xlsx helpers
          req.template.helpers.require = require;
          req.template.helpers.fsproxy = require(_path2.default.join(__dirname, 'fsproxy.js'));
          return _vm2.default.runInNewContext(content, req.template.helpers);
        }

        req.template.helpers = content + '\n' + (req.template.helpers || '');
      });
    });
  });
};
