// jsreport-studio/src/components/Editor/TemplateEditor.js
import React, {Component} from 'react'
import TextEditor from './TextEditor.js'
import 'brace/mode/handlebars'
import 'brace/mode/javascript'
import 'brace/theme/chrome'
import 'brace/ext/searchbox'
import _debounce from 'lodash/debounce'
import SplitPane from '../../components/common/SplitPane/SplitPane.js'
import { triggerSplitResize, templateEditorModeResolvers } from '../../lib/configuration.js'

export default class TemplateEditor extends Component {
  static propTypes = {
    entity: React.PropTypes.object.isRequired,
    onUpdate: React.PropTypes.func.isRequired
  }

  constructor () {
    super()
    this.handleSplitChanged = _debounce(this.handleSplitChanged, 150, { leading: true })
  }

  handleSplitChanged () {
    triggerSplitResize()
  }

  resolveTemplateEditorMode (template) {
    for (const k in templateEditorModeResolvers) {
      const mode = templateEditorModeResolvers[k](template)
      if (mode) {
        return mode
      }
    }

    return null
  }

  componentDidMount () {
    this.refs.contentEditor.ace.editor.focus()
  }

  render () {
    const { entity, onUpdate } = this.props

    return (
      <SplitPane
        split='horizontal' resizerClassName='resizer-horizontal' onChange={() => this.handleSplitChanged()}
        defaultSize={(window.innerHeight * 0.2) + 'px'}>
        <TextEditor
          key={entity._id}
          ref='contentEditor'
          name={entity._id}
          mode={this.resolveTemplateEditorMode(entity) || 'handlebars'}
          onUpdate={(v) => onUpdate(Object.assign({ _id: entity._id }, {content: v}))}
          value={entity.content || ''}
          />
        <TextEditor
          name={entity._id + '_helpers'}
          key={entity._id + '_helpers'}
          mode='javascript'
          onUpdate={(v) => onUpdate(Object.assign({ _id: entity._id }, {helpers: v}))}
          value={entity.helpers || ''}
          />
      </SplitPane>
    )
  }
}
